rasa/shared/core/domain.py

```diff
@@ -352,15 +375,56 @@ def retrieval_intents(self) -> List[Text]:
             if self.intent_properties[intent].get(IS_RETRIEVAL_INTENT_KEY)
         ]
 
+    @classmethod
+    def collect_entity_properties(
+        cls, domain_entities: List[Union[Text, Dict[Text, Any]]]
+    ) -> Tuple[List[Text], Dict[Text, List[Text]], Dict[Text, List[Text]]]:
+        """Get entity properties for a domain from what is provided by a domain file.
+
+        Args:
+            domain_entities: The entities as provided by a domain file.
+
+        Returns:
+            A list of entity names.
+            A dictionary of entity names to roles.
+            A dictionary of entity names to groups.
+        """
+        entities: List[Text] = []
+        roles: Dict[Text, List[Text]] = {}
+        groups: Dict[Text, List[Text]] = {}
+
+        for entity in domain_entities:
+            if isinstance(entity, str):
+                entities.append(entity)
+            elif isinstance(entity, Dict):
+                for _entity, sub_labels in entity.items():
+                    entities.append(_entity)
+                    if ENTITY_ROLES_KEY in sub_labels:
+                        roles[_entity] = sub_labels[ENTITY_ROLES_KEY]
+                    if ENTITY_GROUPS_KEY in sub_labels:
+                        groups[_entity] = sub_labels[ENTITY_GROUPS_KEY]
+            else:
+                raise InvalidDomain(
```

Comment:
when would you expect the type to be something other than `dict` or `str`? shouldn't that be caught in the validation before?

Reply:
Yes, it should be caught before.